src/OpenTelemetry/Metrics/Aggregator/LastValueAggregator.cs

```diff
@@ -23,31 +23,48 @@ namespace OpenTelemetry.Metrics
 {
     public class LastValueAggregator : Aggregator
     {
-        private readonly KeyValuePair<string, object>[] emptyKVP = new KeyValuePair<string, object>[0];
         private readonly Instrument instrument;
-        private readonly string[] names;
-        private readonly object[] values;
+        private readonly KeyValuePair<string, object>[] tags;
 
-        private IDataPoint lastValue;
+        private readonly object lockLastValue = new object();
         private int count = 0;
+        private DataPoint lastDataPoint;
 
         public LastValueAggregator(Instrument instrument, string[] names, object[] values)
         {
+            this.instrument = instrument;
+
             if (names.Length != values.Length)
             {
                 throw new ArgumentException("Length of names[] and values[] must match.");
             }
 
-            this.instrument = instrument;
-            this.names = names;
-            this.values = values;
+            this.tags = new KeyValuePair<string, object>[names.Length];
+            for (int i = 0; i < names.Length; i++)
+            {
+                this.tags[i] = new KeyValuePair<string, object>(names[i], values[i]);
+            }
         }
 
         public override void Update<T>(DateTimeOffset dt, T value)
             where T : struct
         {
-            this.count++;
-            this.lastValue = new DataPoint<T>(dt, value, this.emptyKVP);
+            lock (this.lockLastValue)
+            {
+                this.count++;
+                if (typeof(T) == typeof(int))
```

Comment:
If it's only supported for `int` and `double`, could we make the `LastValueAggregator` class `generic` and  `abstract`.

So something like this: `public abstract class LastValueAggregator<T> : Aggregator` and then have two implementations of it namely:

- `public class LastIntValueAggregator : LastValueAggregator<int>`
- `public class LastDoubleValueAggregator : LastValueAggregator<double>`

with their specific override implementations for `Update<T>`?

Reply:
We will support Byte, Int16, Int32, Int64, Single, Double, and Decimal per .NET API. I just put in int and double to show proof of the initial code structure.

The current design is to have AggregatorStore have one instance of each aggregator type (i.e. Sum, LastValue, Histogram, etc). Each type can handle the individual T via the Update<T> method as it sees fit. This allows for potential "mixing" of T for something like the Histogram.

We can discuss further if you think we need to have T split out earlier in the design, maybe at the AggregatorStore or even the InstrumentState level. I don't currently see an argument for it.

Reply:
The argument here is to keep the `LastValueAggregator` class extensible but closed for modification. I'm not sure how often or if at all the logic for `Update<T>` would change in future but with the current implementation we would have to keep modifying the same `Update<T>` method for any new types that get added or if there is a change in the method logic. Also, doing a runtime check on type in a generic method seems like misusing generics. I believe this can be avoided.

Since the generic type starts at the base class `Aggregator`, we could maybe have the classes defined like the following:
```
public abstract class Aggregator<T> where T : struct
{
    public virtual void Update(DateTimeOffset dt, T value)
    {
        
    }
    ...
}

public abstract class LastValueAggregator<T> : Aggregator<T>
{
    //private fields
    ...
    protected ctor();
    public override IEnumerable<Metric> Collect();
}

public class LastIntValueAggregator : LastValueAggregator<int>
{
    public override void Update(DateTimeOffset dt, int value) {...}
}

public class LastByteValueAggregator : LastValueAggregator<byte> {...}
public class LastShortValueAggregator : LastValueAggregator<short> {...}
public class LastDoubleValueAggregator : LastValueAggregator<double> {...}

```

This would result in a lot of classes but we have more control over making specific modifications as we will now have stricter Single Responsibility classes. Also, another benefit would be that we would avoid the runtime type-checking.

Another benefit of implementing the classes this way would be that we would now not throw an exception if the user were to provide some unsupported type as `T`. We could even enforce them to override `Update` if they decide to use some unsupported type. Although, this a highly unlikely scenario and would imply that the user is wrong here, we should avoid/mitigate this from happening as much as we can since these are public APIs and we have no control on how the user would use them.

Reply:
FYI: I filed #2062 to further discuss this as needed.